File: java/src/localsearchoptimization/components/NeighborhoodWeighted.java
package localsearchoptimization.components;

import java.util.Arrays;

public class NeighborhoodWeighted extends Neighborhood {
    public NeighborhoodWeighted(Solution solution, Operator[] operators, int seed) {
        super(solution, operators, seed);
        double sumOperationsWeight = Arrays.stream(this.operators).mapToDouble(x -> x.weight).sum();
        for (Operator operator : operators) {
            operator.weight /= sumOperationsWeight;
        }
    }

    public Solution getRandom() {
        double rand = random.nextDouble();
        double bound = 0;
        for (Operator operation : operators) {
            if (rand < operation.weight + bound)
                return operation.apply(currentSolution, random.nextInt(operation.power()));
            bound += operation.weight;
        }
        Operator lastOperation = operators[operators.length - 1];
        return lastOperation.apply(currentSolution, random.nextInt(lastOperation.power()));
    }
}

File: C#/LocalSearchOptimization/Components/NeighborhoodWeighted.cs
using System.Collections.Generic;
using System.Linq;

namespace LocalSearchOptimization.Components
{
    public class WeightedNeighborhood : Neighborhood
    {
        public WeightedNeighborhood(ISolution solution, List<Operator> operators, int seed) : base(solution, operators, seed)
        {
            double sumOperationsWeight = this.operators.Sum(x => x.Weight);
            this.operators.ForEach(x =>
            {
                x.Weight = x.Weight / sumOperationsWeight;
            });
        }

        public override ISolution GetRandom()
        {
            double rand = this.random.NextDouble();
            double bound = 0;
            foreach (Operator operation in operators)
            {
                if (rand < operation.Weight + bound)
                    return operation.Apply(CurrentSolution, this.random.Next(operation.Power));
                bound += operation.Weight;
            }
            Operator lastOperation = operators.Last();
            return lastOperation.Apply(CurrentSolution, this.random.Next(lastOperation.Configurations.Count));
        }
    }
}